linked_list/linked_list.py

```diff
@@ -16,113 +16,186 @@ def __init__(self):
     # Time Complexity: ?
     # Space Complexity: ?
     def get_first(self):
-        pass
+        if self.head == None:
+            return None
+        else:
+            return self.head.value
 
 
     # method to add a new node with the specific data value in the linked list
     # insert the new node at the beginning of the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity:O(1)
+    # Space Complexity: O(n)
     def add_first(self, value):
-        pass
+        node = Node(value, next_node = self.head)
+        self.head = node
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: 0
     def search(self, value):
-        pass
+        current_node = self.head
+        while current_node != None:
+            if current_node.value == value:
+                return True
+            current_node = current_node.next
+        return False
+
 
     # method that returns the length of the singly linked list
     # Time Complexity: ?
     # Space Complexity: ?
     def length(self):
-        pass
+        current_node = self.head
+        length = 0
+        while current_node != None:
+            length += 1
+            current_node = current_node.next
+        return length
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
     # Time Complexity: ?
     # Space Complexity: ?
     def get_at_index(self, index):
-        pass
+        current_node = self.head
+        if index >= self.length():
+            return None
+        i = 0
+        while current_node != None:
+            if i == index:
+                return current_node.value
+            i += 1
+            current_node = current_node.next
+        return None
+
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
     # Time Complexity: ?
     # Space Complexity: ?
     def get_last(self):
-        pass
+        current_node = self.head
+        while current_node != None:
+            if current_node.next == None:
+                return current_node.value
+            else: 
+                current_node = current_node.next
+        return None

```

Comment:
This could be a "little" more compact.  This is a *minor* suggestion

```suggestion
        current_node = self.head
        if current_node is None:
            return None
            
        while current_node.next:
            current_node = current_node.next
        
        return current_node
```